src/aws/flb_s3_local_buffer.c

```diff
@@ -414,3 +414,110 @@ int flb_remove_chunk_files(struct flb_local_chunk *c)
     }
     return ret;
 }
+
+/* Generic replace function for strings. */
+char* replace_uri_tokens(const char* original_string, const char* current_word, const char* new_word) 
+{ 
+    char *result; 
+    int i = 0;
+    int count = 0; 
+    int new_word_len = strlen(new_word); 
+    int old_word_len = strlen(current_word); 
+  
+    for (i = 0; original_string[i] != '\0'; i++) { 
+        if (strstr(&original_string[i], current_word) == &original_string[i]) { 
+            count++; 
+            i += old_word_len - 1; 
+        } 
+    }
+
+    result = flb_sds_create_size(i + count * (new_word_len - old_word_len) + 1);
+    if (!result) {
+        flb_errno();
+        return NULL;
+    }
+
+    i = 0; 
+    while (*original_string) { 
+        if (strstr(original_string, current_word) == original_string) { 
+            strcpy(&result[i], new_word); 
+            i += new_word_len; 
+            original_string += old_word_len; 
+        } 
+        else
+            result[i++] = *original_string++; 
+    } 
+
+    result[i] = '\0'; 
+    return result; 
+} 
+  
+//* Constructs S3 object key as per the format. */
+flb_sds_t get_s3_key(const char *format, time_t time, char *tag, char *tag_delimiter)
+{
+    int i = 0;
+    int max_key_len = (strlen(format) + strlen(tag) + 1);
```

Comment:
this calculation is definitely not accurate... remember that the tag can occur multiple times, and that also the strftime formatters could add an unknown number of extra characters. 

There are a few solutions to a problem like this:
- dynamically allocate memory at each step when you know how much is needed (basically, as you build the key figure out how long it actually is).
- Use the flb_sds_t library and its functions, since the whole point of it is that it will re-allocate memory for strings as needed. Its supposed to be a "string type" more like what you'd be used to in higher level languages
- Use some "guess", like setting len to  5 * format + 5 * tag. It'd be extremely unlikely for the final key to be longer than that. This is a okay technique only if you are constantly checking if the real len has not actually exceeded the guess. Of these options, you should see that this one is the convenient.

Reply:
Hmm okay, I read further down and see how you are actually using this... see my comment below..

Reply:
You are right the max_len here will fail. Since max filename length of an s3 object key is 1024 characters. We can set that as our max length and throw a warning if it exceeds that limit. What would be your opinion on this approach?

Reply:
That works. You could allocate a 1024 SDS.

Or. You could put the memory on the stack: `char buf[1024]`. And then find the actual length with `strlen(buf)`. And then copy it into an SDS. That'd be slightly more efficient. I don't think it really matters though.

Reply:
Added a check to ensure the object key is no longer than 1024.

Reply:
Checks are done

- at the beginning to ensure format is lesser than 1024
- after every specific tag replace e.g.$TAG[*]
- after every $TAG replace
- after time is formatted.

We need the above 4 checks because our final return string s3_key has been declared with a max size of 1024. 
`s3_key = flb_sds_create_len(format, S3_KEY_SIZE);`

If at any point the string assigned to s3_key is longer than 1024 then by default only the first 1024 characters are copied (no error will be thrown). So it is important that we check at these 4 places so that an appropriate error message can be displayed.